src/main/lrsql/init/authority.clj

```diff
@@ -33,23 +34,32 @@
    :cred-id       #uuid "00000000-0000-4000-0000-000000000001"
    :account-id    #uuid "00000000-0000-4000-0000-000000000002"})
 
-(defn- valid-authority-fn?
-  [authority-fn]
-  ;; When adding basic authorities (i.e. NOT 3-legged OAuth),
-  ;; the authority MUST be an Agent.
-  (s/valid? ::xs/agent (authority-fn sample-authority-fn-input)))
+(defn validate-authority-fn*
+  "Returns `authority-fn` if it's valid, throws an exception if
+   `(authority-fn sample-auth-fn-input)` does not satisfy `:statement/authority`,
+   and logs `warn-msg` if it does not satisfy `warn-spec`."
+  [authority-fn template-path warn-spec sample-auth-fn-input warn-msg]
+  (let [sample-authority (authority-fn sample-auth-fn-input)]
+    (when-not (s/valid? :statement/authority sample-authority)
+      (throw (ex-info "Authority template does not produce a valid xAPI Authority."
+                      {:type          ::invalid-json
+                       :template-path template-path})))
+    (when-not (s/valid? warn-spec sample-authority)
+      (log/warn warn-msg))
```

Comment:
Yep, I like this approach. Maybe put in a TODO to remove when we move to xAPI 2.0.0/IEEE 9274

Reply:
https://github.com/yetanalytics/lrsql/pull/248/commits/b211f152bb4f2f7c980852b91af3383727cdc2d2